connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java

```diff
@@ -269,7 +289,7 @@ void backoff(int attempt, long deadline) {
             delay = deadline - time.milliseconds();
         }
         log.debug("Sleeping for {} millis", delay);
-        time.sleep(delay);
+        return !exitLatch.await(delay, TimeUnit.MILLISECONDS);
```

Comment:
I don't think we should use a return value here to indicate whether it's safe to continue or not, since it's unclear how to handle interruptions (if we get interrupted before we're scheduled for shutdown, we should keep retrying; if we're interrupted after being scheduled for shutdown, we should stop).

We can use the pattern from `AbstractWorkerSourceTask` in this class: use a latch for backing off, but also set a boolean flag when we're scheduled for shutdown, and use the latter when checking whether we should perform another retry iteration.

The control flow I have in mind is something like this:
- `backoff` tries to block the thread for the specified delay, but can return early due to interruption, shutdown, whatever. There is no return value, and it swallows `InterruptedException` instances by just returning normally
- We check the boolean flag before performing any retries and, if it's set, preemptively exit
- `exit` counts down the latch and sets the boolean flag to indicate that we should halt execution

Reply:
I think I get what you're saying - but why can't we simply catch `InterruptedException` here in this method itself and return `true` indicating that backoff can be called again? Then if shutdown hasn't been scheduled yet - i.e. the latch hasn't counted down to 0 yet, the next backoff will be normal; but if shutdown has been scheduled - the next call to `ShutDownLatch::await` will return immediately and `backoff()` will return false resulting in no further retries.

Reply:
With that approach, we would retry the operation one more time before invoking `backoff`.

Reply:
Ah, that's a very good point! I didn't account for this because I had (incorrectly) assumed that if a thread's interrupt status is set and a `CountDownLatch` has already been counted down to 0, a call to `CountDownLatch::await` would return immediately with `true` (i.e. the latch's count would be given "priority" over the thread's interrupt status). However, this isn't actually the case and the call to `CountDownLatch::await` will instead immediately throw an `InterruptedException` 😄